src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,478 @@
+        let circle_outer_offset_y  = Var::<Distance<Pixels>>::from(circle_outer_offset_y);
+        let circle_outer           = circle_outer.translate_y(circle_outer_offset_y);
+
+        let circle_inner_offset_y = &shape_height - segment_outer.sagitta();
+        let circle_inner_offset_y = circle_inner_offset_y - segment_inner.sagitta() + &inner_radius;
+        let circle_inner_offset_y = Var::<Distance<Pixels>>::from(circle_inner_offset_y);
+        let circle_inner          = circle_inner.translate_y(circle_inner_offset_y);
+
+        // Now we shape the angle by applying the circles.Note that only one of them will have an
+        // effect, based on the radius that we modified set through `is_inwards`.
+        let sculpted_shape = base_shape;
+        let sculpted_shape = Intersection(sculpted_shape,circle_outer);
+        let sculpted_shape = Difference(sculpted_shape,circle_inner);
+        let sculpted_shape = sculpted_shape.fill(Srgb::new(0.26, 0.69, 0.99));
+
+        // The angle should be centered on (0,0) to make it easier to rotate and minimise the
+        // required canvas.
+        let center_offset    = Var::<Distance<Pixels>>::from(&shape_height * Var::from(0.5));
+        let sculpted_shape   = sculpted_shape.translate_y(-center_offset);
+
+        // This is a conditional rotation that allows the port to either point inwards
+        // or outwards.
+        let pi             = Var::from(PI);
+        let rotation_angle = is_inwards * pi;
+        let rotation_angle = Var::<Angle<Radians>>::from(rotation_angle);
+        let sculpted_shape = sculpted_shape.rotate(rotation_angle);
+
+        sculpted_shape.into()
+    }
+
+    ensogl::define_shape_system! {
+        (height:f32,width:f32,inner_radius:f32,is_inwards:f32) {
+            // FIXME: `is_inwards` should only be 0.0 or 1.0.
+            new_port(height,width,inner_radius,is_inwards)
+        }
+    }
+
+    impl Shape{
+        /// Set the shape parameters derived from the `Specification`.
+        pub fn update_from_spec(&self,spec:&Specification){
+            self.update_parameters(spec.height,spec.inner_radius, spec.width, spec.direction)
+        }
+
+        /// Set the shape parameters.
+        pub fn update_parameters
+        (&self, height:f32, inner_radius:f32, width:Angle<Degrees>, direction:Direction){
+            self.height.set(height);
+            self.inner_radius.set(inner_radius);
+            self.width.set(width.value.to_radians());
+            match direction{
+                Direction::In  => self.is_inwards.set(1.0),
+                Direction::Out => self.is_inwards.set(0.0),
+            };
+        }
+    }
+}
+
+
+
+// ============
+// === Port ===
+// ============
+
+/// Initialise the shape with sensible defaults..
+fn init_shape(shape:&shape::Shape, direction:Direction){
+    let spec  = Specification::default();
+    shape.update_from_spec(&spec);
```

Comment:
I dont follow it really. We are creating a new port shape. When creating the port it has all parameters already set, like height. Then we call this function, which then calls `update_parameters`, which then sets the parameters again. What is the purpose of that?

Reply:
Actually this is the correct place to initialise the shape. The initialisation in the constructor doesn't work if the shape is not accessible (which it isn't at the moment). I've removed the initialisation of the shape there.